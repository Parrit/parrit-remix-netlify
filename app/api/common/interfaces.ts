export interface LoginRequest {
  projectName?: string;
  password?: string;
}

export interface LoginResult {
  message: string;
}

export interface Project {
  id: string;
  name: string;
  pairingBoards: PairingBoard[];
  people: Person[];
}

export interface PairingBoard {
  id: string;
  name: string;
  exempt: boolean;
  people: Person[];
  roles: Role[];
}

export interface Person {
  id: string;
  name: string;
}

export interface Role {
  id: string;
  name: string;
}

export interface PairingArrangementDTO {
  id: number; // this should be changed to string
  pairingTime: string;
  pairingHistories: PairingHistory[];
}

export interface PairingHistory {
  pairingBoardName: string;
  people: Person[];
}
