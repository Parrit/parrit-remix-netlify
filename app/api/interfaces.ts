export interface FormErrors<T> {
  server?: string;
  fields?: Record<keyof T, string | null>;
  status: 400 | 401 | 500;
}

export const isFormErrors = <T>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  arg: FormErrors<T> | any
): arg is FormErrors<T> => {
  return (arg as FormErrors<T>).status > 0;
};
