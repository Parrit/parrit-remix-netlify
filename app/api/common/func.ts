import {
  FLOATING_IDX,
  PairingBoard,
  Person,
  Project,
} from "./interfaces/parrit.interfaces";
import * as R from "ramda";

// Functional programing utilities for working with interfaces

export const movePerson = (
  project: Project,
  person: Person,
  position: PairingBoard
) =>
  R.pipe<[Project], Project, Project>(
    removePerson(person),
    addPerson(person, position)
  )(project);

export const addPerson = R.curry(
  (person: Person, position: PairingBoard, project: Project) => {
    const f1 = addPersonToPosition(person, position);
    const f2 =
      position.id === FLOATING_IDX ? addPersonToFloating(person) : R.identity;
    return R.pipe<[Project], Project, Project>(f1, f2)(project);
  }
);

const addPersonToPosition = R.curry(
  (person: Person, position: PairingBoard, project: Project) =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      R.map<PairingBoard, PairingBoard>((x) => {
        if (x.id === position.id) {
          return R.over<PairingBoard, Person[]>(
            R.lensProp("people"),
            R.append({ ...person, pairing_board_id: position.id }),
            x
          );
        }
        return x;
      }),
      project
    )
);

const addPersonToFloating = R.curry((person: Person, project: Project) =>
  R.over<Project, Person[]>(
    R.lensPath(["floating", "people"]),
    R.append(person),
    project
  )
);

export const removePerson = R.curry(
  (person: Person, project: Project): Project =>
    R.pipe<[Project], Project, Project>(
      removePersonFromBoards(person),
      removePersonFromFloating(person)
    )(project)
);

const removePersonFromBoards = R.curry(
  (person: Person, project: Project): Project =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      R.map<PairingBoard, PairingBoard>((x) => {
        if (x.people.includes(person)) {
          return R.set<PairingBoard, Person[]>(
            R.lensProp("people"),
            R.without(x.people, [person]),
            x
          );
        }
        return x;
      }),
      project
    )
);

const removePersonFromFloating = R.curry(
  (person: Person, project: Project): Project =>
    R.set<Project, Person[]>(
      R.lensPath(["floating", "people"]),
      R.without([person], project.floating.people),
      project
    )
);

const pbIndexById = R.curry((pbs: PairingBoard[], pb_id: string) =>
  R.findIndex<PairingBoard>((pb) => pb.id === pb_id)(pbs)
);

const renamePairingBoardFromList = R.curry(
  (pb_id: string, rename: string, pbs: PairingBoard[]) => {
    const foundIndex = pbIndexById(pbs, pb_id);
    if (foundIndex < 0) {
      return pbs;
    }
    return R.adjust<PairingBoard>(foundIndex, R.assoc("name", rename), pbs);
  }
);

export const renamePairingBoard = R.curry(
  (project: Project, pb_id: string, rename: string): Project =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      renamePairingBoardFromList(pb_id, rename),
      project
    )
);

export const canAPairingBeMade = (project: Project): boolean => {
  const atLeast2Floaters = project.floating.people.length >= 2;
  const atLeast1Floater = project.floating.people.length >= 1;
  const atLeast1EmptyBoard = getEmptyPairingBoard(project) !== undefined;
  const atLeast1UnpairedSticker = unpairedStickingPeople(project).length >= 1;
  return (
    (atLeast1Floater && atLeast1UnpairedSticker) ||
    (atLeast2Floaters && atLeast1EmptyBoard)
  );
};

export const getEmptyPairingBoard = (
  project: Project
): PairingBoard | undefined => {
  return project.pairingBoards.find(
    (pb) => pb.people.length === 0 && !pb.exempt
  );
};

const unpairedStickingPeople = (project: Project): Person[] =>
  project.pairingBoards
    .filter((board) => !board.exempt)
    .filter((board) => board.people.length === 1)
    .flatMap((board) => board.people);

const findPairingBoardByPerson = (
  project: Project,
  person: Person
): PairingBoard | undefined =>
  project.pairingBoards.find((pb) => pb.people.find((p) => p.id === person.id));
