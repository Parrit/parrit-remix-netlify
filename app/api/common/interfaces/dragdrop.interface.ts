export type DragType = "Person" | "Role";

export enum DropType {
  Floating = "Floating",
  PairingBoard = "PairingBoard",
  TrashBin = "TrashBin",
}

export interface DragItem {
  id: string;
  name: string;
  type: DragType;
  pairing_board_id: string;
}

export interface DropItem {
  type: DropType;
}

export interface DropTarget {
  isOver: boolean;
  canDrop: boolean;
}

export interface DragResult {}
