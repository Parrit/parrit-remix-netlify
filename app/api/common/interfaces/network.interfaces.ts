export interface LoginRequest {
  projectName?: string;
  password?: string;
}

export interface LoginResult {
  message: string;
}
