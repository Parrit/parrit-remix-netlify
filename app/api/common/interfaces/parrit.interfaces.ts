import { DragItem } from "~/api/common/interfaces/dragdrop.interface";

export interface Project {
  id: string;
  name: string;
  pairingBoards: PairingBoard[];
  floating: PairingBoard;
  people: Person[]; // the new cannonical location for people
  roles: Role[]; // the new cannonical location for roles
}

export const FLOATING_IDX = "-1";

export interface PairingBoard {
  id: string;
  name: string;
  exempt: boolean;
}

export interface Person extends DragItem {
  id: string;
  name: string;
  type: "Person";
}

export interface Role extends DragItem {
  id: string;
  name: string;
  type: "Role";
}

export interface ProjectPairingSnapshot {
  id: string;
  projectId: string;
  pairingTime: string;
  pairingInstances: PairingInstance[];
}

export interface PairingInstance {
  projectId: string;
  pairingBoardName: string | null;
  people: Person[];
  pairingTime: string;
}
