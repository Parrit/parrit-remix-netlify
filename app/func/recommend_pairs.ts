import {
  Project,
  Person,
  PairingInstance,
  FLOATING_IDX,
} from "~/api/common/interfaces/parrit.interfaces";
import {
  can_a_pairing_be_made,
  floating_people,
  get_empty_pairing_board,
  unpaired_sticking_people,
} from "./utils";
import move_person from "./move_person";

export function recommendPairs(project: Project, history: PairingInstance[]) {
  const helper = new ProjectHelper(project, history);
  return helper.recommendedConfiguration();
}

class Pair {
  private person1;
  private person2;

  constructor(person1?: Person, person2?: Person) {
    // guarantee that this pair will be formed the same no matter the order
    // that the people are input
    if (!person1 || !person2) {
      console.error(person1, person2);
      throw new Error("Pair must be initialized with two people");
    }
    const arr = [person1, person2].sort((a, b) => a.id.localeCompare(b.id));
    this.person1 = arr[0];
    this.person2 = arr[1];
  }

  get hashcode() {
    return `${this.person1.id}&&${this.person2.id}`;
  }
}

export class ProjectHelper {
  private project: Project;
  timetable: { [key: string]: Date };

  constructor(project: Project, history: PairingInstance[]) {
    this.project = project;
    this.timetable = {};
    history.forEach((item) => {
      if (item.people.length > 1) {
        for (let i = 0; i < item.people.length; i++) {
          for (let j = i + 1; j < item.people.length; j++) {
            const p1 = item.people[i];
            const p2 = item.people[j];
            try {
              const pair = new Pair(p1, p2);
              const hash = pair.hashcode;
              const extant = this.timetable[hash];
              const pairDate = new Date(item.pairingTime);
              if (!extant) {
                this.timetable[hash] = pairDate; // create date from timestamp
              } else {
                if (extant.getTime() < pairDate.getTime()) {
                  this.timetable[hash] = pairDate;
                }
              }
            } catch (e) {
              console.error(e);
            }
          }
        }
      }
    });
  }

  recommendedConfiguration(): Project {
    let projectCopy = { ...this.project };
    while (can_a_pairing_be_made(projectCopy)) {
      projectCopy = this.iterateMatch(projectCopy);
    }
    this.project = projectCopy;
    return this.project;
  }

  iterateMatch(project: Project) {
    const floatingPerson = floating_people(project).at(0);
    if (!floatingPerson) {
      console.warn("attempting no floating people, a match cannot be made");
      return project;
    }
    let topPair = this.pairFor(floatingPerson, project);
    for (let i = 0; topPair != undefined; i++) {
      topPair = this.pairFor(floatingPerson, project, i);
      if (topPair) {
        const targetPairingBoard = topPair.pairing_board_id;
        if (targetPairingBoard !== FLOATING_IDX) {
          const proj = move_person(project, floatingPerson, targetPairingBoard);
          return proj;
        } else {
          // we know top pair is floating because they have no pairing board
          const emptyPairingBoard = get_empty_pairing_board(project);
          if (emptyPairingBoard && !emptyPairingBoard.exempt) {
            let proj = move_person(
              project,
              floatingPerson,
              emptyPairingBoard.id
            );
            proj = move_person(proj, topPair, emptyPairingBoard.id);
            return proj;
          } else {
            // found 2 unmatched pairs but no empty pairing board
          }
        }
      }
    }

    return project;
  }

  pairFor(
    person?: Person,
    project: Project = this.project,
    nth: number = 0
  ): Person | undefined {
    if (!person) {
      console.warn("attempting to find pair for nonexistant person");
      return undefined;
    }
    // 0-indexed number from the last most recent paired
    const allAvailable = [
      ...floating_people(project),
      ...unpaired_sticking_people(project),
    ].filter((p) => !!p && p.id !== person.id);
    const partnerDates = allAvailable
      .map((p) => {
        if (person.id === p.id) {
          return undefined;
        }
        const hashcode = new Pair(person, p).hashcode;
        return { time: this.timetable[hashcode], partner: p };
      })
      .filter((el) => {
        return el !== undefined;
      })
      .sort((a, b) => {
        if (!a?.time) {
          // this pairing has never occured
          return -1;
        }
        if (!b?.time) {
          // this pairing has never occured
          return 1;
        }
        return a.time.getTime() - b.time.getTime();
      });
    if (partnerDates.length > 0) {
      return partnerDates[nth]?.partner;
    }

    return undefined;
  }
}
