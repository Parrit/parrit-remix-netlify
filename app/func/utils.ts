/**
 * These files are tested via the functionality of other tests.
 * That's what makes them utils and not their own functions
 */

import {
  Project,
  Person,
  PairingBoard,
} from "~/api/common/interfaces/parrit.interfaces";

export const can_a_pairing_be_made = (project: Project): boolean => {
  const atLeast2Floaters = project.floating.people.length >= 2;
  const atLeast1Floater = project.floating.people.length >= 1;
  const atLeast1EmptyBoard = get_empty_pairing_board(project) !== undefined;
  const atLeast1UnpairedSticker = unpaired_sticking_people(project).length >= 1;
  return (
    (atLeast1Floater && atLeast1UnpairedSticker) ||
    (atLeast2Floaters && atLeast1EmptyBoard)
  );
};

export const unpaired_sticking_people = (project: Project): Person[] =>
  project.pairingBoards
    .filter((board) => !board.exempt)
    .filter((board) => board.people.length === 1)
    .flatMap((board) => board.people);

export const get_empty_pairing_board = (
  project: Project
): PairingBoard | undefined => {
  return project.pairingBoards.find(
    (pb) => pb.people.length === 0 && !pb.exempt
  );
};

export const all_people_in_project = (project: Project): Person[] => {
  const people: Person[] = [...project.floating.people];
  project.pairingBoards.forEach((pb) =>
    pb.people.forEach((person) => people.push(person))
  );
  return people;
};

export const find_pairing_board_by_person = (
  project: Project,
  person: Person
): PairingBoard | undefined =>
  project.pairingBoards.find((pb) => pb.people.find((p) => p.id === person.id));
