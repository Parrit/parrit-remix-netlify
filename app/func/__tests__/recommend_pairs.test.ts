import { Person } from "~/api/common/interfaces/parrit.interfaces";
import { recommendPairs } from "../recommend_pairs";
import { can_a_pairing_be_made } from "../utils";
import { PROJECT_START } from "./common";
import { describe, expect, it } from '@jest/globals';

describe("recommend_pairs", () => {
  it("recommends pairs", () => {
    expect(can_a_pairing_be_made(PROJECT_START)).toBe(true);
    const after = recommendPairs(PROJECT_START, []);
    expect(after).toMatchSnapshot();
  });

  it("doesn't shove everyone onto the same board", () => {
    const before = {
      ...PROJECT_START,
      people: [
        ...PROJECT_START.people,
        ...[1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map(
          (num) =>
            ({
              id: `floaty_${num}`,
              name: `Floaty${num}`,
              pairing_board_id: "-1",
              type: "Person",
            } as Person)
        ),
      ],
    };
    const after = recommendPairs(before, []);
    expect(after).toMatchSnapshot();
  });
});
