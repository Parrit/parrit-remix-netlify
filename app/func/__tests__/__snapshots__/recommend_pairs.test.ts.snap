// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recommend_pairs doesn't shove everyone onto the same board 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "floating_board_name",
  },
  "id": "project_id",
  "name": "project_name",
  "pairingBoards": [
    {
      "exempt": false,
      "id": "red_id",
      "name": "red_start",
    },
    {
      "exempt": false,
      "id": "blue_id",
      "name": "blue_start",
    },
    {
      "exempt": true,
      "id": "exempt",
      "name": "out_of_office",
    },
  ],
  "people": [
    {
      "id": "red_parrit_id",
      "name": "red_parrit",
      "pairing_board_id": "red_id",
      "type": "Person",
    },
    {
      "id": "absent",
      "name": "Permanent Vacation",
      "pairing_board_id": "exempt",
      "type": "Person",
    },
    {
      "id": "floaty_2",
      "name": "Floaty2",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "floaty_3",
      "name": "Floaty3",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "floaty_4",
      "name": "Floaty4",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "floaty_5",
      "name": "Floaty5",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "floaty_6",
      "name": "Floaty6",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "floaty_7",
      "name": "Floaty7",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "floaty_8",
      "name": "Floaty8",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "floaty_9",
      "name": "Floaty9",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "floater_id",
      "name": "floater_name_start",
      "pairing_board_id": "red_id",
      "type": "Person",
    },
    {
      "id": "floaty_1",
      "name": "Floaty1",
      "pairing_board_id": "blue_id",
      "type": "Person",
    },
    {
      "id": "floaty_10",
      "name": "Floaty10",
      "pairing_board_id": "blue_id",
      "type": "Person",
    },
  ],
  "roles": [
    {
      "id": "blue_role",
      "name": "Blue Role",
      "pairing_board_id": "blue_id",
      "type": "Role",
    },
    {
      "id": "ooo",
      "name": "OOOROLE",
      "pairing_board_id": "exempt",
      "type": "Role",
    },
  ],
}
`;

exports[`recommend_pairs recommends pairs 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "floating_board_name",
  },
  "id": "project_id",
  "name": "project_name",
  "pairingBoards": [
    {
      "exempt": false,
      "id": "red_id",
      "name": "red_start",
    },
    {
      "exempt": false,
      "id": "blue_id",
      "name": "blue_start",
    },
    {
      "exempt": true,
      "id": "exempt",
      "name": "out_of_office",
    },
  ],
  "people": [
    {
      "id": "red_parrit_id",
      "name": "red_parrit",
      "pairing_board_id": "red_id",
      "type": "Person",
    },
    {
      "id": "absent",
      "name": "Permanent Vacation",
      "pairing_board_id": "exempt",
      "type": "Person",
    },
    {
      "id": "floater_id",
      "name": "floater_name_start",
      "pairing_board_id": "red_id",
      "type": "Person",
    },
  ],
  "roles": [
    {
      "id": "blue_role",
      "name": "Blue Role",
      "pairing_board_id": "blue_id",
      "type": "Role",
    },
    {
      "id": "ooo",
      "name": "OOOROLE",
      "pairing_board_id": "exempt",
      "type": "Role",
    },
  ],
}
`;
